functions/functions.go

```diff
@@ -0,0 +1,157 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package functions provides objective functions for testing optimization
+// algorithms.
+//
+// We encourage outside contributions of additional test functions that exhibit
+// properties not already covered in the testing suite or that have
+// significance due to prior use as benchmark cases.
+package functions
+
+import (
+	"math"
+	"testing"
+
+	"github.com/gonum/diff/fd"
+	"github.com/gonum/floats"
+)
+
+// Function represents an objective function.
+type function interface {
+	F(x []float64) float64
+}
+
+type gradient interface {
+	Df(x, grad []float64)
+}
+
+type functionGradient interface {
+	FDf(x, grad []float64) float64
+}
+
+// minimumer is an objective function that can also provide information about
+// its minima.
+type minimumer interface {
+	function
+
+	// Minima returns _known_ minima of the function.
+	Minima() []Minimum
+}
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. X may not be nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates if the location is a global minimum.
+	Global bool
+}
+
+type funcTest struct {
+	X []float64
+
+	// F is the expected function value at X.
+	F float64
+	// Gradient is the expected gradient at X. If nil, it is not evaluated.
+	Gradient []float64
+}
+
+// TODO: Decide and implement an exported testing function:
+// func Test(f Function, ??? ) ??? {
+// }
+
+const (
+	defaultTol       = 1e-12
+	defaultGradTol   = 1e-9
+	defaultFDGradTol = 1e-5
+)
+
+// testFunction checks that the function can evaluate itself (and its gradient)
+// correctly.
+func testFunction(f function, ftests []funcTest, t *testing.T) {
+	// Make a copy of tests because we may append to the slice.
+	// TODO: is this necessary?
```

Comment:
Just take off this TODO. A better solution isn't obvious to me, and I'm not worried about the memory allocation.